[BUG] in `ConformalIntervals`, fix update of residuals matrix for sliding window splitter

bethrice44 opened the pull request:
#### Reference Issues/PRs
Fixes https://github.com/sktime/sktime/issues/3913


#### What does this implement/fix? Explain your changes.
Ensures only indices in the new residuals matrix are populated if already calculated and avoids the KeyError.


#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?


#### Any other comments?


#### PR checklist

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.


<!--
Thanks for contributing!
-->

fkiraly reviewed (changes requested):
looks plausible! - may I request to cover this by a test?

It might be as simple as adding a parameter set with the sliding window splitter in `get_test_params`.

bethrice44 commented:
Thanks @fkiraly  will add a test.

While thinking about the test I realised a limitation of the `_update` method. It is set up to not recalculate `residuals_matrix` rows which have already been calculated. However, because the additional rows will add new columns, you end up with the existing rows having NULL in the last x columns. This has the potential to cause issues if for example the window cv is long and you are building a lot of the matrix via the `_update` method. 

Does the problem make sense? 

I have fixed by defining an `_extend_residuals_matrix` function which uses the already fitted forecasters and extends the prediction to the length of the new `y_test`.

I have added a test to check `ConformalIntervals` and `NaiveVariance` with strategy = `refit` or `update` and `cv` =  sliding or expanding. The test threw up an erroneous assertion error in `PinballLoss` which I have also added a fix for (we use `convert_dtypes()` from pandas which converts to `Float64` instead of `float`).

GuzalBulatova commented on sktime/performance_metrics/forecasting/probabilistic/_classes.py:
```diff
@@ -256,7 +256,7 @@ def _check_consistent_input(self, y_true, y_pred, multioutput):
             raise ValueError("y_pred should be a dataframe.")
 
-        if not all(y_pred.dtypes == float):
+        if not all(y_pred.dtypes == float) and not all(y_pred.dtypes == "Float64"):
```
Should it be "or" here?

bethrice44 replied:
That's what I thought initially but because it is `not` it has to be `and` (I think). I tested both and `or` didn't give expected behaviour. So it only raises the error if all of the dtypes are NOT float AND all of the dtypes are NOT Float64 (OR with NOTs give weird behaviour because it will always be true that they are NOT one or the other).

fkiraly replied:
I think it should be `not any(y_pred.dtypes not in [float, Float64])`, or the condition will erroneously fire if some columns are `Float64` and the remaining ones are `float`.

An easier way, I think, is using `is_numeric_dtype`?

bethrice44 replied:
Ok, that makes sense - will use `is_numeric_dtype`

GuzalBulatova commented on sktime/forecasting/tests/test_interval_wrappers.py:
```diff
@@ -4,15 +4,24 @@
 
 __author__ = ["fkiraly"]
```
Please add your GitHub handle here too

GuzalBulatova reviewed (commented):
Only have a couple of small comments. Great work, @bethrice44!

fkiraly reviewed (approved):
Thanks, @bethrice44!

Re your questions, your handling of extending the residuals matrix look sensible to me.

At some point, we may like to factor this out more cleanly, as it is common to multiple predictive probabilistic wrappers. Would appreciate a design, we can also talk about that together with other devs.
